pymaker/gas.py

```diff
@@ -86,121 +107,160 @@ def get_gas_price(self, time_elapsed: int) -> Optional[int]:
         without an additional HTTP request.  This baseclass exists to let a subclass manipulate the node price."""
         raise NotImplementedError("Please implement this method")
 
-    def get_node_gas_price(self):
-        return max(self.web3.manager.request_blocking("eth_gasPrice", []), 1 * self.GWEI)
+    def get_gas_fees(self, time_elapsed: int) -> Optional[Tuple[int, int]]:
+        """Implementation of tip is subjective.  For August 2021, the following implementation is a reasonable example:
+        return int(self.get_next_base_fee(self)*1.5), 2 * self.GWEI"""
+        raise NotImplementedError("Please implement this method")
+
+    def get_node_gas_price(self) -> int:
+        return max(self.web3.manager.request_blocking("eth_gasPrice", []), 1)
+
+    def get_base_fee(self) -> Optional[int]:
+        """Useful for calculating maxfee; a multiple of this value is suggested"""
+        pending_block = self.web3.eth.get_block('pending')
+        if 'baseFeePerGas' in pending_block:
+            return max(int(pending_block['baseFeePerGas']), 1)
+        else:
+            return None
 
 
-class FixedGasPrice(GasPrice):
+class FixedGasPrice(GasStrategy):
     """Fixed gas price.
 
     Uses specified gas price instead of the default price suggested by the Ethereum
     node the keeper is connected to. The gas price may be later changed (while the transaction
     is still in progress) by calling the `update_gas_price` method.
 
     Attributes:
-        gas_price: Gas price to be used (in Wei).
+        gas_price: Gas price to be used (in Wei) for legacy transactions
+        max_fee:   Maximum fee (in Wei) for EIP-1559 transactions, should be >= (base_fee + tip)
+        tip:       Priority fee (in Wei) for EIP-1559 transactions
     """
-    def __init__(self, gas_price: int):
-        assert(isinstance(gas_price, int))
+    def __init__(self, gas_price: Optional[int], max_fee: Optional[int] = None, tip: Optional[int] = None):
+        assert isinstance(gas_price, int) or gas_price is None
+        assert isinstance(max_fee, int) or max_fee is None
+        assert isinstance(tip, int) or tip is None
+        assert gas_price or (max_fee and tip)
```

Comment:
Is it a good idea to check exclusive or between types of transactions?

Reply:
The idea with this line is that the user doesn't _need_ to speciffy a type 0 `gas_price` if they populate type 2 fields.  On the right of the `or` we're confirming that if they do specify type 2 fields, they must specify both.  This isn't required for EIP-1559, but it is (erroneously) required for web3, which raises an error if both aren't specified.